src/Compiler/Facilities/AsyncMemoize.fs

```diff
@@ -1,19 +1,85 @@
 namespace Internal.Utilities.Collections
 
 open System
-open System.Collections.Generic
 open System.Diagnostics
 open System.IO
 open System.Threading
 open System.Threading.Tasks
 
-open FSharp.Compiler
-open FSharp.Compiler.BuildGraph
-open FSharp.Compiler.Diagnostics
 open FSharp.Compiler.DiagnosticsLogger
 open Internal.Utilities.Library
 open System.Runtime.CompilerServices
 
+type AsyncLazyState<'t> =
+    | Initial of Async<'t>
+    | Created of Task<'t> * CancellationTokenSource * int
```

Comment:
```suggestion
    | Created of Task<'t> * CancellationTokenSource * requestCount: int
```

For clarity.

Reply:
Also, I would prefer `Started` instead of `Created`. 

Might be worth to also add `Finished`/`Completed`? 🤔

Reply:
I thought about it. My reasoning is that this evokes the lazy semantics (the value is either created or not), while the actual running state of the job is already kept by the task (TaskStatus, IsCompleted and so on).
So this just delegates that part of the state machine to the Task, where this is already implemented.

Alternative implementation would look like this probably:
```fsharp
type AsyncLazyState<'t> =
    | Initial of Async<'t>
    | Started of TaskCompletionSource<'t> * CancellationTokenSource * requestCount: int
    | Finished of TaskCompletionSource<'t>
```
Closer to the original, maybe more clearer, too.

Reply:
Or maybe even call it `Running`.

Reply:
Yeah it's technically redundant but in a few places we're "querying" for it so this might make it clearer. Though it adds the extra state transition. 

I thought it could be useful if we decide to restart failed jobs. 

But definitely not a must.

Reply:
I'll attempt later to do this with TaskCompletionSource and Async.StartWithContinuations. Unlike StartAsTask it runs immediate. If it works, we could make the first request on the same thread.